pkg/ingester/ingester_v2.go

```diff
@@ -455,3 +488,93 @@ func (i *Ingester) closeAllTSDB() {
 	i.userStatesMtx.Unlock()
 	wg.Wait()
 }
+
+/* cleanup walks the TSDB directory for each user to determine if the TSDB should be closed.
+ * If
+ * - The directory is empty of blocks
+ * OR
+ * - The directory has every block marked as shipped
+ * AND
+ * - The TSDB hasn't been written to within the alloted time period
+ * THEN
+ * - Close the TSDB, stop the shipper routine, and delete the directory and it's contents.
+ */
+func (i *Ingester) cleanup() error {
+	ts := time.Now()
+	defer func() {
+		i.metrics.cleanupDuration.Observe(time.Since(ts).Seconds())
+	}()
+
+	userIDs := i.getUsers()
+
+	unshipped, err := unshippedBlocks(i.cfg.TSDBConfig.Dir)
+	if err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	for _, user := range userIDs {
+
+		// Only try and process users that don't have unshipped blocks
+		if len(unshipped[user]) != 0 {
+			continue
+		}
+
+		// Get the last written time stamp
+		udb := i.getTSDB(user)
+		if udb == nil {
+			level.Warn(util.Logger).Log("msg", "unable to find user tsdb during cleanup", "user", user)
+			continue
+		}
+
+		udb.Lock()
+		last := udb.lastWritten
+		udb.Unlock()
+
+		// If the TSDB is inactive, close and remove it.
+		if ts.Sub(last) > i.cfg.TSDBConfig.MaxStaleAge {
+			wg.Add(1)
+			go func(user string) {
+				defer wg.Done()
+				i.userStatesMtx.Lock()
+				defer i.userStatesMtx.Unlock()
+
+				db, ok := i.TSDBState.dbs[user]
+				if ok {
+
+					// Stop the shipper routine for this TSDB
+					db.cancel()
+
+					// Close the TSDB
+					if err := db.Close(); err != nil {
```

Comment:
Closing the TSDB, any samples written in the head but not yet cut into a new block will not be uploaded to the storage. We need to ensure a new block from the head is cut and a last successful run of the ship is completed before moving on (especially before deleting data on disk).

Reply:
Yup that's a good catch. I'll refactor some things to do this. I think it'll be a two pass operation.

Reply:
Okay, I've refactored some of that code, have it take a snapshot before closing. I think this does a better job of covering corner cases.